### ISSUE TITLE
Loss behaviour in eval_batch

### ISSUE THREAD

**Issue by NickHauptvogel** [deephealthproject/pyeddl#61]
Hi,

I am currently trying to train a Convolutional Neural Network (namely ["StackedHourglass"](https://arxiv.org/abs/1603.06937)) in pyeddl-gpu version 0.14.0 on a NVIDIA GTX 960.

My first training runs, however, brought up some issues.

1. **Explosion of loss**
While executing the eddl.train_batch() method, the loss is decreasing steadily as it should. However, in eddl.eval_batch(), the loss sometimes suddenly explodes, leading to high loss values which are taken up by the next eddl.train_batch() run, worsening the model more and more. I cannot explain why the loss explodes and worsens the next training loop, as eval_batch() should not change the model's weights.
Do I miss something? Do I have to call eddl.reset_loss() when using eddl.train_batch() and eddl.eval_batch()? What does eddl.reset_loss() do anyway, isn't a loss only volatile, not stored in a model and only computed by forward propagation?

The output of the training log (Loss explosion in epoch 3):

```
-------------- NETWORK ARCHITECTURE --------------

INFO:root:WIDTH: 256
INFO:root:HEIGHT: 256
INFO:root:INPUT CHANNELS: 1
INFO:root:OUTPUT CHANNELS: 14
INFO:root:LOSSES: ['mse']
INFO:root:METRICS: ['mse']
...
-------------- END NETWORK ARCHITECTURE --------------

CS with full memory setup
INFO:root:
-------------- TRAINING CONFIGURATION --------------

INFO:root:EPOCHS 40
INFO:root:BATCH SIZE 4
INFO:root:INITIAL LR 0.01
...
-------------- END TRAINING CONFIGURATION --------------

Building model
Selecting GPU device 0
EDDLL is running on GPU device 0, GeForce GTX 960
CuBlas initialized on GPU device 0, GeForce GTX 960
CuRand initialized on GPU device 0, GeForce GTX 960
INFO:root:
============== TRAINING ==============

INFO:root:
-------------- EPOCH 1/40 --------------

INFO:root:Executing train_batch (2625 batches)
INFO:root:Batch: 1 Loss: [13.097932815551758] Metric: [13.097932815551758]
INFO:root:Batch: 51 Loss: [12.502132415771484] Metric: [12.502132415771484]
INFO:root:Batch: 101 Loss: [12.473625183105469] Metric: [12.473625183105469]
INFO:root:Batch: 151 Loss: [12.463068008422852] Metric: [12.463068008422852]
INFO:root:Batch: 201 Loss: [12.457537651062012] Metric: [12.457537651062012]
INFO:root:Batch: 251 Loss: [12.45468807220459] Metric: [12.45468807220459]
...
INFO:root:Executing eval_batch (749 batches)
INFO:root:Batch: 1 Loss: [12.413102149963379] Metric: [12.413102149963379]
INFO:root:Batch: 51 Loss: [12.413569450378418] Metric: [12.413569450378418]
...
-------------- EPOCH 3/40 --------------

INFO:root:Executing train_batch (2625 batches)
INFO:root:Batch: 1 Loss: [13.944581031799316] Metric: [13.944581031799316]
INFO:root:Batch: 51 Loss: [13.933237075805664] Metric: [13.933237075805664]
INFO:root:Batch: 101 Loss: [13.922091484069824] Metric: [13.922091484069824]
INFO:root:Batch: 151 Loss: [13.911008834838867] Metric: [13.911008834838867]
...
INFO:root:Executing eval_batch (749 batches)
INFO:root:Batch: 1 Loss: [13.51264762878418] Metric: [13.51264762878418]
INFO:root:Batch: 51 Loss: [13.506909370422363] Metric: [13.506909370422363]
INFO:root:Batch: 101 Loss: [13.501219749450684] Metric: [13.501219749450684]
INFO:root:Batch: 151 Loss: [13.495572090148926] Metric: [13.495572090148926]
INFO:root:Batch: 201 Loss: [13.489946365356445] Metric: [13.489946365356445]
INFO:root:Batch: 251 Loss: [13.484417915344238] Metric: [13.484417915344238]
INFO:root:Batch: 301 Loss: [13.478986740112305] Metric: [13.478986740112305]
INFO:root:Batch: 351 Loss: [1496636544.0] Metric: [1496636544.0]
INFO:root:Batch: 401 Loss: [1488994560.0] Metric: [1488994560.0]
INFO:root:Batch: 451 Loss: [1481416192.0] Metric: [1481416192.0]
...
-------------- EPOCH 4/40 --------------

INFO:root:Executing train_batch (2625 batches)
INFO:root:Batch: 1 Loss: [1911438336.0] Metric: [1911438336.0]
INFO:root:Batch: 51 Loss: [1902043648.0] Metric: [1902043648.0]
INFO:root:Batch: 101 Loss: [1892740864.0] Metric: [1892740864.0]
INFO:root:Batch: 151 Loss: [1883528704.0] Metric: [1883528704.0]

```

The code extract for the training is shown here:
```python
    def train(self):
            ...
            for i in range(training_config.get_epochs()):
                logging.info(f'\n-------------- EPOCH {i + 1}/{training_config.get_epochs()} --------------\n')

                training_loss, training_metric = self.__execute_per_data_set(self.train_data,
                                                                             self.train_labels,
                                                                             model_config,
                                                                             eddl.train_batch,
                                                                             batch_size=training_config.get_batch_size())

                val_loss, val_metric = self.__execute_per_data_set(self.val_data,
                                                                   self.val_labels,
                                                                   model_config,
                                                                   eddl.eval_batch,
                                                                   batch_size=training_config.get_batch_size())
            ...

    def __execute_per_data_set(self, x, y, model_config, method, batch_size):

        x_size = len(x)
        checkpoint = max(x_size // 10, 1)
        num_batches = x_size // batch_size
        logging.info(f'Executing {method.__name__} ({num_batches} batches)')

        losses = []
        metrics = []
        for j in range(num_batches):

            x_tensor = Tensor.fromarray(x[j*batch_size:(j+1)*batch_size])
            y_tensor = Tensor.fromarray(y[j*batch_size:(j+1)*batch_size])

            method(model_config.get_built_model(), [x_tensor], [y_tensor] * len(model_config.get_losses()))
            loss1 = eddl.get_losses(model_config.get_built_model())
            metric1 = eddl.get_metrics(model_config.get_built_model())
            losses.append(sum(loss1) / len(loss1))
            metrics.append(sum(metric1) / len(metric1))
            if j % checkpoint == 0:
                logging.info(f'Batch: {j + 1} Loss: {loss1} Metric: {metric1}')

        return sum(losses) / len(losses), sum(metrics) / len(metrics)
```

2. Evaluate loss inconsistency
Inspired by this realization, I performed some further tests in my training loop. For example, I repeatedly called eddl.eval_batch() to see whether the results match each other, which they should (with the same data of course). Comparing eddl.eval_batch() to eddl.evaluate() I found that loss scores from sequential calls differ on eddl.eval_batch(). In evaluate(), they are always the same. Here, I always evaluated the model on a small data set consisting of 10 samples.

Log:
```
-------------- EPOCH 1/3 --------------

INFO:root:Executing train_batch (3 batches)
INFO:root:Batch: 1 Loss: [13.08705997467041] Metric: [13.08705997467041]
INFO:root:Batch: 2 Loss: [12.91485595703125] Metric: [12.91485595703125]
INFO:root:Batch: 3 Loss: [12.829597473144531] Metric: [12.829597473144531]

1. RUN
INFO:root:eval_batch Loss: [43.422332763671875] Metric: [43.422332763671875]
INFO:root:evaluate Loss: [61.83473587036133] Metric: [61.8347282409668]

2. RUN (evaluate() loss matches to before, eval_batch() does not)
INFO:root:eval_batch Loss: [61.77797317504883] Metric: [61.77797317504883]
INFO:root:evaluate Loss: [61.8347282409668] Metric: [61.8347282409668]
```

Code:
```python
    def train(self):
            ...
            for i in range(training_config.get_epochs()):
                logging.info(f'\n-------------- EPOCH {i + 1}/{training_config.get_epochs()} --------------\n')

                eddl.reset_loss(model_config.get_built_model())

                training_loss, training_metric = self.__execute_per_data_set(self.train_data,
                                                                             self.train_labels,
                                                                             model_config,
                                                                             eddl.train_batch,
                                                                             batch_size=training_config.get_batch_size())

                self.__validate_example_set(model_config)
                eddl.reset_loss(model_config.get_built_model())
                self.__validate_example_set(model_config)
                ...

    def __validate_example_set(self, model_config):

        x_train = Tensor.load("x_train", "bin")
        y_train = Tensor.load("y_train", "bin")

        eddl.eval_batch(model_config.get_built_model(), [x_train], [y_train])
        loss1 = eddl.get_losses(model_config.get_built_model())
        metric1 = eddl.get_metrics(model_config.get_built_model())
        logging.info(f'eval_batch Loss: {loss1} Metric: {metric1}')

        eddl.evaluate(model_config.get_built_model(), [x_train], [y_train], 1)
        loss1 = eddl.get_losses(model_config.get_built_model())
        metric1 = eddl.get_metrics(model_config.get_built_model())
        logging.info(f'evaluate Loss: {loss1} Metric: {metric1}')
```

I feel like I am missing something in combination with eval_batch().

For reconstruction,  I will send @salvacarrion the .onnx model and the binary files which were used for validation.

Thank you and kind regards,

Nick

**Comment by simleo** [comment]
This is not about the bindings. Hopefully the EDDL team can help. Also tagging @giobus75, who makes heavy use of the library and might have some ideas / advice.

**Comment by giobus75** [comment]
Hi, for historical reasons I never used the eval_batch function but I perform the evaluation calling a set of other APIs. As you can see [here](https://github.com/deephealthproject/promort_pipeline/blob/master/python/promort_cassandra.py) starting at line 251, after loading sample and target tensors, I call the forward API function and, after that, I loop across each sample output to compute the loss average over the entire batch. I would suggest trying this approach to understand if the problem is related to the `eval_batch` function. 
Hope it helps
Giovanni

**Comment by salvacarrion** [comment]
I'll need to replicate this model in c++ to debug it. Thanks!

**Comment by NickHauptvogel** [comment]
Okay, I think I have made some progess in my investigations.

First, implementing the `reset_loss` function at the start of each epoch seems to prevent the high loss to be taken up into the next training loop. I talked with @salvacarrion and we worked out that this is necessary in order to reset the loss which is saved in the net.

Regarding the difference of `evaluate`, `eval_batch` and the custom implementation by @giobus75  with `forward` for the validation: I make the assumption that `evaluate` and the custom impl calculate the loss for each batch alone, whereas `eval_batch` computes the average if the loss is not reset in between. This explains why the loss, once exploded, was kept at a high level in the log file above.

However, one problem remains: After switching to `evaluate`, I could investigate the loss value of each validation batch.  ~99% of loss values are fine and acceptable, but ~1% generate a very high loss (as far as 10E15 or more). Examining the predictions visually, the model is predicting complete non-sense in those pictures, and I do not know why.

I will open another issue for that.

Kind regards and thanks for your help,

Nick
